client/client.go

```diff
@@ -141,9 +155,81 @@ func NewClient(baseDir string) (*Client, error) {
 
 // Currently only closes the history file and shuts down the sync thread.
 func (c *Client) Close() error {
+	if testMode {
+		// Dump the logs & events to the test client directory.
+		path := filepath.Join(c.staticBaseDir, "status.txt")
+		os.WriteFile(path, []byte(c.DumpEventLogs()), 0644)
+	}
 	return c.tg.Stop()
 }
 
+// latestLogWithPrefix searches for the most recent logs starting with a prefix,
+// and prints it with a newline.
+func latestLogWithPrefix(logs map[string][]time.Time, prefix string) string {
+
+	var t time.Time
+	var line string
+	for key, times := range logs {
+		// If the latest time is past t, this is latest.
+		if strings.HasPrefix(key, prefix) && times[len(times)-1].After(t) {
+			t = times[len(times)-1]
+			line = key
+		}
+	}
+	if len(line) == 0 {
+		return ""
+	}
+	return t.UTC().Format(time.RFC3339) + " " + line + "\n"
+}
+
+// Helper to dump client status to a string. Returns general information,
+// followed by the event log.
+func (c *Client) DumpEventLogs() string {
+	var sb strings.Builder
+
+	now := time.Now()
+
+	sb.WriteString(fmt.Sprintf("UTC:    %v\n", now.UTC().Format(time.RFC3339)))
+	sb.WriteString(fmt.Sprintf("Local:  %v\n", now.Format(time.RFC3339)))
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
```

Comment:
I ultimately decided this was fine, but generally speaking you want to avoid holding a lot while doing a ton of string processing - it's actually one of the slowest things you can do in go. Nothing in the client is really going to have issues if a mutex is held for 20 milliseconds to process a giant log (the risky part is the loop), but in other codebases (perhaps even the server code, where there's a user facing API) I generally wouldn't allow a mutex to be held during a loop that could be iterating hundreds of thousands of times and also doing string processing each iteration.